### PR TITLE
[Core] Pipeline Parallel support for Model Runner V2

### PR DESCRIPTION
**Pull request by ZhanqiuHu**
## Summary

Co-authored with @yewentao256

Add Pipeline Parallel (PP) support to Model Runner V2 (`vllm/v1/worker/gpu/model_runner.py`). This introduces a modular `PPHandler` class that encapsulates all PP logic, keeping the model runner code clean. Verified correct output and competitive throughput against the V1 baseline.

Related: #32455 (Q1 2026 Roadmap) — PP is listed as a missing feature for Model Runner V2.

### In Model Runner (`vllm/v1/worker/gpu/model_runner.py`)

- **`execute_model`**:
  - First rank: Run model forward with raw token inputs, send `IntermediateTensors` to next stage.
  - Middle ranks: Accept `IntermediateTensors` from previous stage, run model forward, send `IntermediateTensors` to next stage.
  - Last rank: Accept `IntermediateTensors` from previous stage, run model forward, store hidden states for sampling.
  - Delegates input/output preparation to `PPHandler`.

- **`sample_tokens`**:
  - Last rank: Sample tokens, broadcast `sampled_token_ids`/`num_sampled`/`num_rejected` to all other ranks, then return `ModelRunnerOutput`.
  - Non-last ranks: Receive broadcast tensors, call `postprocess` to update local state, return `None`.

- **`_dummy_run`**: Create dummy intermediate tensors for non-first ranks; skip sampler for non-last ranks.
- **`capture_model`**: Skip CUDA graph capture when PP is enabled (eager-only for now).
- **Deterministic request sorting**: Use `req_id` as tie-breaker to ensure consistent batch ordering across PP ranks.
- **Multimodal guard**: Only prepare MM embeddings on the first PP rank.

### New Class: PPHandler (`vllm/v1/worker/gpu/pp_handler.py`)

New module with a `PPHandler` class. All public methods no-op when PP is disabled or called on an inapplicable rank, so callers don't need guard conditions.

| Method | Description |
|--------|-------------|
| `maybe_broadcast_sampled_tokens` | Last rank broadcasts `sampled_token_ids`, `num_sampled`, `num_rejected` to all PP ranks. No-ops on non-last ranks. |
| `maybe_receive_sampled_tokens` | Non-last ranks receive the broadcast tensors. Returns `None` on last rank. Supports variable `max_sample_len` for future speculative decoding + PP. |
| `prepare_model_inputs` | Builds the `model.forward()` kwargs — raw inputs for first rank, intermediate tensors for others. |
| `prepare_output` | Extracts hidden states (last rank) or wraps as `IntermediateTensors` (non-last ranks). |

Helper classes: `PPConfig` (dataclass with rank role, size, index) and `PPRole` enum (`NO_PP`, `FIRST`, `MIDDLE`, `LAST`).

## Future Work

- [ ] CUDA graph support for PP (currently eager-only)
- [ ] Async (non-blocking) next-sampled-token transfer
- [ ] Async intermediate tensor send/recv
- [ ] Speculative decoding + PP integration

## Test Plan

**Accuracy** — lm_eval gsm8k (5-shot):
```bash
export MODEL="Qwen/Qwen3-30B-A3B-Thinking-2507-FP8"

# V1 baseline
vllm serve $MODEL -pp 2 --port 9256 --enable-expert-parallel --max-num-seqs 128 --enforce-eager

# V2
VLLM_USE_V2_MODEL_RUNNER=1 vllm serve $MODEL -pp 2 --port 9256 --enable-expert-parallel --max-num-seqs 128 --enforce-eager

# Evaluate
lm_eval --model local-completions \
  --model_args "base_url=http://127.0.0.1:9256/v1/completions,model=$MODEL,num_concurrent=1024" \
  --tasks gsm8k
```

**Throughput** — decode-heavy random workload:
```bash
vllm bench serve --model $MODEL --dataset-name random --host 127.0.0.1 \
  --random-input-len 2 --random-output-len 512 --num-prompts 128 \
  --port 9256 --num-warmups 16
```

### Accuracy (gsm8k, PP=2, Qwen3-30B-A3B MoE FP8)

| Metric | V1 | V2 |
|--------|----|----|
| flexible-extract (exact_match) | 0.6641 ± 0.0130 | 0.6641 ± 0.0130 |
| strict-match (exact_match) | 0.7801 ± 0.0114 | 0.7794 ± 0.0114 |

**V2 matches V1 accuracy.** The 0.0007 difference in strict-match is within noise (±0.0114 stderr).

### Throughput (PP=2, Qwen3-30B-A3B MoE FP8, input=2, output=512, 128 prompts)

| Metric | V1 | V2 | Diff |
|--------|----|----|------|
| Output token throughput (tok/s) | 7315.42 | 6956.99 | -4.9% |
| Total token throughput (tok/s) | 7344.00 | 6984.17 | -4.9% |
| Mean TTFT (ms) | 170.33 | 194.20 | +14.0% |
| Mean TPOT (ms) | 17.12 | 17.99 | +5.1% |
| Mean ITL (ms) | 17.12 | 17.99 | +5.1% |

---

*Note: Claude (Anthropic) was used as a coding assistant during development of this PR.*

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by mengjian0502** [comment]
🚀🚀🚀

File: vllm/v1/worker/gpu/model_runner.py [hunk 1]
```diff
@@ -889,6 +935,15 @@ def execute_model(
...
                 assert input_batch.mrope_positions is not None
                 positions = input_batch.mrope_positions
+
+            # PP input preparation: handler centralizes all PP input logic.
+            model_inputs = self.pp_handler.prepare_model_inputs(
+                input_batch.input_ids,
+                positions,
+                input_batch.inputs_embeds,
+                intermediate_tensors,
+            )
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py]: Can we keep the model inputs explicit? I’m not a fan of encapsulating them into a separate object. For example, when supporting CUDA graphs, it’s critical to reason about all inputs and ensure they use consistent memory addresses. This abstraction makes that harder to see.

File: vllm/v1/worker/gpu/model_runner.py [hunk 2]
```diff
@@ -899,27 +954,52 @@ def execute_model(
...
-                    inputs_embeds=input_batch.inputs_embeds,
-                )
+                hidden_states = self.model(**model_inputs)
 
+        # PP output handling: use handler so all PP output logic lives in one place.
         kv_connector_output = self.kv_connector.post_forward(scheduler_output)
-        self.execute_model_state = hidden_states, input_batch, kv_connector_output
+        output = self.pp_handler.prepare_output(hidden_states, kv_connector_output)
+
+        if isinstance(output, IntermediateTensors):
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py]: Can we just check the PP rank directly?

File: vllm/v1/worker/gpu/model_runner.py [hunk 3]
```diff
@@ -899,27 +954,52 @@ def execute_model(
...
-                    input_ids=input_batch.input_ids,
-                    positions=positions,
-                    inputs_embeds=input_batch.inputs_embeds,
-                )
+                hidden_states = self.model(**model_inputs)
 
+        # PP output handling: use handler so all PP output logic lives in one place.
         kv_connector_output = self.kv_connector.post_forward(scheduler_output)
-        self.execute_model_state = hidden_states, input_batch, kv_connector_output
+        output = self.pp_handler.prepare_output(hidden_states, kv_connector_output)
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py]: Same. It'd be nicer to make `output` here more explicit.

File: vllm/v1/worker/gpu/model_runner.py [hunk 4]
```diff
@@ -303,13 +312,31 @@ def _dummy_run(
         # Disable any use of KVConnector for dummy runs.
         self.kv_connector.set_disabled(True)
 
+        # For non-first PP ranks, create dummy intermediate_tensors.
+        intermediate_tensors = None
+        if not self.pp_handler.receives_raw_inputs:
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py]: Can we just use `is_first_rank` directly?

File: vllm/v1/worker/gpu/model_runner.py [hunk 5]
```diff
@@ -175,6 +181,10 @@ def __init__(
         # KV Connector if configured.
         self.kv_connector: KVConnector = NO_OP_KV_CONNECTOR
 
+        # PP Handler for pipeline parallelism.
+        # Initialize here since distributed groups are set up before model runner.
+        self.pp_handler: PPHandler = get_pp_handler(self.parallel_config)
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py]: For better readability, can we not initialize `pp_handler` at all when PP is not used?
As PP is not a super popular feature, I think we should prioritize users/developers who do not want to be bothered by it.

File: vllm/v1/worker/gpu/model_runner.py [hunk 6]
```diff
@@ -54,6 +56,10 @@
 from vllm.v1.worker.gpu.lora_utils import LoraState
 from vllm.v1.worker.gpu.mm.encoder_runner import EncoderRunner
 from vllm.v1.worker.gpu.mm.mrope_utils import MRopeState
+from vllm.v1.worker.gpu.pp_handler import (
+    PPHandler,
+    get_pp_handler,
+)
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py]:
```suggestion
from vllm.v1.worker.gpu.pp_handler import PPHandler,mget_pp_handler
```

File: vllm/v1/worker/gpu/model_runner.py [hunk 7]
```diff
@@ -3,19 +3,21 @@
...
 import numpy as np
 import torch
 import torch.nn as nn
 
 from vllm.config import VllmConfig
 from vllm.config.compilation import CUDAGraphMode
-from vllm.distributed.parallel_state import prepare_communication_buffer_for_model
+from vllm.distributed.parallel_state import (
+    prepare_communication_buffer_for_model,
+)
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py:16]: revert?

File: vllm/v1/worker/gpu/model_runner.py [hunk 8]
```diff
@@ -899,27 +954,52 @@ def execute_model(
...
 
+        # Non-last PP rank: receive tokens and update state.
+        # Don't need to sample tokens here - just receive from last rank.
+        if hidden_states is None:
+            received = self.pp_handler.maybe_receive_sampled_tokens(
+                input_batch.num_reqs,
+                self.device,
+                max_sample_len=self.num_speculative_steps + 1,
+            )
+            if received is not None:
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py:991]: When is it None?

File: vllm/v1/worker/gpu/model_runner.py [hunk 9]
```diff
@@ -899,27 +954,52 @@ def execute_model(
...
+
+        # Last rank: sample tokens
         sampler_output, num_sampled, num_rejected = self.sample(
             hidden_states, input_batch, grammar_output
         )
+
+        # Broadcast to non-last ranks (handles spec decode multi-token)
+        self.pp_handler.maybe_broadcast_sampled_tokens(
+            sampler_output, num_sampled, num_rejected
+        )
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py]: I think it'd be nice if we can do something like
```suggestion
        if self.use_pp:
            # Broadcast to non-last ranks (handles spec decode multi-token)
            self.pp_handler.maybe_broadcast_sampled_tokens(
                sampler_output, num_sampled, num_rejected
            )
```

File: vllm/v1/worker/gpu/model_runner.py [hunk 10]
```diff
@@ -303,13 +312,31 @@ def _dummy_run(
...
+            dummy_scheduler_output,
+            intermediate_tensors=intermediate_tensors,
+            dummy_run=True,
+            skip_attn_for_dummy_run=skip_attn,
         )
         self.kv_connector.set_disabled(False)
+
+        # Non-last PP ranks don't produce output for sampling.
+        if not self.pp_handler.produces_final_output:
+            return None, None
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py:336]: Let's use `is_last_rank` directly

**Review by WoosukKwon — CHANGES_REQUESTED** [review]
@ZhanqiuHu Thanks for the PR! I like the design of using PP handler to encapsulate the PP-related logics.

That said, I think we could improve the model runner code and make the change even smaller. Please check out my comments.

File: vllm/v1/worker/gpu/model_runner.py [hunk 11]
```diff
@@ -506,7 +545,15 @@ def prepare_inputs(
...
-        req_ids = sorted(num_tokens_per_req, key=num_tokens_per_req.get)  # type: ignore[arg-type]
+        # NOTE: In PP mode, every rank must construct the *exact* same request
+        # ordering for the batched token dimension. Python's `sorted(..., key=...)`
+        # is stable, so ties would otherwise be broken by the input dict's
+        # insertion order, which can differ across processes. Use `req_id` as a
+        # deterministic tie-breaker to keep PP stages in sync.
+        req_ids = sorted(
+            num_tokens_per_req,
+            key=lambda req_id: (num_tokens_per_req[req_id], req_id),
+        )
```
Comment (njhill) [inline vllm/v1/worker/gpu/model_runner.py]: I don't think this should be needed. The `num_tokens_per_req` dict received by each rank should be identical, and `sorted` is stable.

I found that this sort is actually quite a bit faster without the lambda.

Comment (ZhanqiuHu) [reply vllm/v1/worker/gpu/model_runner.py]: Reverted

File: vllm/v1/worker/gpu/model_runner.py [hunk 12]
```diff
@@ -899,27 +940,71 @@ def execute_model(
...
+                        positions=positions,
+                        inputs_embeds=None,
+                        intermediate_tensors=intermediate_tensors,
+                    )
+                else:
+                    hidden_states = self.model(
+                        input_ids=input_batch.input_ids,
+                        positions=positions,
+                        inputs_embeds=input_batch.inputs_embeds,
+                    )
```
Comment (njhill) [inline vllm/v1/worker/gpu/model_runner.py:957]: suggested:
```suggestion
                if get_pp_group().is_first_rank:
                    input_ids = input_batch.input_ids
                    inputs_embeds = input_batch.inputs_embeds
                else:
                    # Non-first PP rank: forward with intermediate tensors.
                    input_ids, inputs_embeds = None, None
                    assert intermediate_tensors is not None

                hidden_states = self.model(
                    input_ids=input_ids,
                    positions=positions,
                    inputs_embeds=inputs_embeds,
                    intermediate_tensors=intermediate_tensors,
                )
```

File: vllm/v1/worker/gpu/model_runner.py (hunk 8 above)
Comment (njhill) [reply vllm/v1/worker/gpu/model_runner.py:991]: @ZhanqiuHu ^^

**Review by WoosukKwon — APPROVED** [review]
@ZhanqiuHu LGTM! Thanks for addressing all comments. I'm excited that we have such a clean implementation of PP. Great work 👍
